source/Calamari/Kubernetes/Commands/KubernetesDiscoveryCommand.cs

```diff
@@ -0,0 +1,172 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Calamari.Commands.Support;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Discovery;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.ServiceMessages;
+using Calamari.Common.Plumbing.Variables;
+using Calamari.Kubernetes.Commands.Discovery;
+using JetBrains.Annotations;
+using Newtonsoft.Json;
+
+namespace Calamari.Kubernetes.Commands
+{
+    [Command(Name, Description = "Discovery Kubernetes cluster targets")]
+    [UsedImplicitly]
+    public class KubernetesDiscoveryCommand : Command
+    {
+        public const string Name = "kubernetes-target-discovery";
+        public const string CreateKubernetesTargetServiceMessageName = "create-kubernetestarget";
+        public const string ContextVariableName = "Octopus.TargetDiscovery.Context";
+        
+        readonly ILog log;
+        readonly IVariables variables;
+        readonly IKubernetesDiscovererFactory discovererFactory;
+
+        public KubernetesDiscoveryCommand(ILog log,
+            IVariables variables,
+            IKubernetesDiscovererFactory discovererFactory)
+        {
+            this.log = log;
+            this.variables = variables;
+            this.discovererFactory = discovererFactory;
+        }
+
+        /// <returns>
+        /// The Discovery Command always returns 0 to indicate success
+        /// because a failure to discover targets should not cause a
+        /// deployment process to fail.
+        /// </returns>
+        public override int Execute(string[] commandLineArguments)
+        {
+            if (!TryGetDiscoveryContextJson(out var json))
+            {
+                log.Warn($"Could not find target discovery context in variable {ContextVariableName}.");
+                return 0;
+            }
+            
+            if (!TryGetAuthenticationContextTypeAndDiscoveryContextScope(json, out var type, out var scope))
+                return 0;
```

Comment:
@gb-8 I spoke to you about this briefly, are you happy for the discovery command to always return 0 (success) even if it encounters an error like the Discovery Context json being malformed?

Reply:
If this can only fail due to defects, then I think we can fail the process here.
When there is a failure due to user misconfiguration, or cloud outage, etc. then it should return 0 so the deployment can proceed.

Reply:
Yeah I believe in this situation, the json must be correct unless there is a code issue or server is producing badly shaped DiscoveryContext json. I think I can change it to a non-zero return code for this case then.